knodle/transformation/majority.py

```diff
@@ -100,115 +129,273 @@ def z_t_matrices_to_majority_vote_probs(
+        mapping_rules_labels_t: Mapping of rules to labels, binary encoded. Shape: rules x classes
+        choose_random_label_for_ties: Whether a random label is chosen, if there's no clear majority vote.
+        choose_other_label_for_ties: Whether an "other" label is chosen, if there's no clear majority vote.
+        other_class_id: the id of other class, i.e. the class of negative samples
+    Returns: Decision per sample. Shape: (instances, )
+    """
+    rule_counts_probs = z_t_matrices_to_probs_multi(rule_matches_z, mapping_rules_labels_t)
+
+    kwargs = {
+        "choose_random_label": choose_random_label_for_ties,
+        "choose_other_label": choose_other_label_for_ties,
+        "other_class_id": other_class_id
+    }
+
+    majority_labels = np.zeros((rule_counts_probs.shape[0], rule_counts_probs.shape[1], 1))
+    for i, prob_row in enumerate(rule_counts_probs):
+        mv_per_token = np.apply_along_axis(probabilities_to_majority_vote, axis=1, arr=prob_row, **kwargs)
+        majority_labels[i] = np.reshape(mv_per_token, (-1, 1))
+    return majority_labels
+
+
+def z_t_matrices_to_probs_multi(
+        rule_matches_z: np.ndarray,
+        mapping_rules_labels_t: np.ndarray,
+        normalization: str = "softmax"
+) -> np.ndarray:
+    """
+    This function calculates a majority vote probability for all rule_matches_z. The difference from simple
+    get_majority_vote_probs function is the following: samples, where no rules matched (that is, all elements in
+    the corresponding raw in rule_matches_z matrix equal 0), are assigned to no_match_class (that is, a value in the
+    corresponding column in rule_counts_probs matrix is changed to 1).
+    Args:
+        rule_matches_z: Binary encoded array of which rules matched. Shape: instances x rules
+        mapping_rules_labels_t: Mapping of rules to labels, binary encoded. Shape: rules x classes
+        other_class_id: Class which is chosen, if no function is hitting.
+        normalization: The way how the vectors will be normalized. Currently, there are two supported normalizations:
+            - softmax
+            - sigmoid
+    Returns: Array with majority vote probabilities. Shape: instances x classes
+    """
+    rule_probs = np.zeros((rule_matches_z.shape[0], rule_matches_z.shape[1], mapping_rules_labels_t.shape[1]))
+    for i, inst_row in enumerate(rule_matches_z):
+        if inst_row.shape[1] != mapping_rules_labels_t.shape[0]:
+            raise ValueError(f"Dimensions mismatch! Z matrix has shape {rule_matches_z.shape}, while "
+                             f"T matrix has shape {mapping_rules_labels_t.shape}")
+
+        if isinstance(inst_row, sp.csr_matrix):
+            rule_counts = inst_row.dot(mapping_rules_labels_t)
+            if isinstance(rule_counts, sp.csr_matrix):
+                rule_counts = rule_counts.toarray()
+        else:
+            rule_counts = np.matmul(inst_row, mapping_rules_labels_t)
+
+        if normalization == "softmax":
```

Comment:
this is not softmax but sort of "standard". Not sure about name, but softmax is wrong, would need exp(.)